lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,277 @@
+/// and using the compilation context \p cctx.
+/// \returns constant results.
+std::vector<Constant *>
+evaluateConstantOperation(Backend &backend, CompilationContext &cctx, Node *C) {
+  PlaceholderBindings bindings;
+  assert(isConstantOperation(C, backend) && "Expected a constant expression");
+  // Constants and splats do not need to be constant evaluated.
+  if (isa<Constant>(C) || isa<SplatNode>(C)) {
+    return {};
+  }
+  Module &mod = *C->getParent()->getParent();
+  // Create a temporary function to perform the constant operation.
+  Function *constEvaluationF = mod.createFunction(constEvaluationFunctionName);
+  // Mapping from existing nodes to the new ones.
+  NodeMap currToNew;
+  // Clone the constant operation and some of its inputs if necessary.
+  auto *clonedC = recursiveClone(constEvaluationF, C, currToNew);
+  // Create save nodes for each of the results.
+  llvm::SmallVector<SaveNode *, 16> savedResults;
+  for (unsigned idx = 0, e = clonedC->getNumResults(); idx < e; ++idx) {
+    auto *SN = constEvaluationF->createSave("save", clonedC->getNthResult(idx));
+    savedResults.emplace_back(SN);
+    bindings.allocate(SN->getPlaceholder());
+  }
+  // Run the temporary backend to perform this constant operation
+  // evaluation.
+  EXIT_ON_ERR(executeFunction(backend, *constEvaluationF, bindings, cctx,
+                              /* isConstant */ true));
+  // Get the results of the constant operation compile-time computation and
+  // create new constants from it.
+  std::vector<Constant *> constResults;
+  constResults.reserve(savedResults.size());
+  for (auto *SN : savedResults) {
+    Tensor *outputTensor = bindings.get(SN->getPlaceholder());
+    auto *constResult =
+        mod.createConstant(strFormat("constant"), *outputTensor);
+    constResults.emplace_back(constResult);
+  }
+  // Remove the temporary function.
+  mod.eraseFunction(constEvaluationF);
+  return constResults;
+}
+
+/// Check if function \p F consists of constant operations only.
+llvm::Error verifyConstantFunction(Backend &backend, Function &F) {
+  for (auto &N : F.getNodes()) {
+    // Saving results is fine.
+    if (isa<SaveNode>(&N)) {
+      continue;
+    }
+    // Placeholders can be used just to save results.
+    if (isa<Placeholder>(&N)) {
+      if (N.hasOneUse()) {
+        auto SN = dyn_cast<SaveNode>(N.getUsers().begin()->getUser());
+        if (SN && SN->getPlaceholder() == &N) {
+          continue;
+        }
+      }
+      RETURN_ERR("Expected constant operation");
+    }
+    RETURN_ERR_IF_NOT(isConstantOperation(&N, backend),
+                      "Expected constant operation");
+  }
+  return llvm::Error::success();
+}
+
+} // namespace
+
+llvm::Error glow::executeFunction(Backend &backend, Function &F,
+                                  PlaceholderBindings &bindings,
+                                  CompilationContext &cctx, bool isConstant) {
+  if (isConstant) {
```

Comment:
So the motivation here of only returning this error if `isConstant` is if there's an external caller that doesn't expect to use this with a Constant expression? Do you have an expected use case for the future in this way?

If so I'd suggest moving it somewhere other than `ConstantFolding.cpp` as it's kind of confusing, maybe even to a different library so it can be reused without depending on the GraphOptimizer, e.g. somewhere in the runtime.

If not then I'd suggest removing `isConstant` and rename it to `executeConstantFunction()` and always return an error in this case.

Reply:
Removed `isConstant` for now.